src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -179,6 +189,15 @@ export class WorkspaceSavedObjectsClientWrapper {
     return hasPermission;
   }
 
+  // Data source is a workspace level object, validate if data source has the permission to the given `workspaceIds`
+  private validateDataSourcePermissions = (
+    object: SavedObject,
+    request: OpenSearchDashboardsRequest
+  ) => {
+    const requestWorkspaceId = getWorkspaceState(request).requestWorkspaceId;
+    return !requestWorkspaceId || !!object.workspaces?.includes(requestWorkspaceId);
```

Comment:
`!requestWorkspaceId` this covers the case when outside of a workspace?

Reply:
No, in `WorkspaceState` type, `requestWorkspaceId` could be undefined. Here just want to ensure that the workspaceId to be compared is existing.

Reply:
The `requestWorkspaceId` should be exists if user in a workspace. In another words,  user should be out of workspace if requestWorkspaceId is undefined right?

Reply:
We don't need to care whether is in or out of workspace especially here.

Reply:
Imaging user is in global(requestWorkspaceId is undefined) and try to get a datasource(without workspaces) by `id`, we should prevent user from getting the datasource detail.

Reply:
I think this may be a common case in OSD, this doesn't belong to the concept of workspace/assign/unassign, we should not do extra limitation.

Reply:
@wanglam @ruanyl what's your opinion?
IMO, regarding the restriction will only be turned on when permission control and workspace is enabled, `Reading` it from all kinds of unexpected approaches should be prevented.

Reply:
Yeah, agreed. There may be some use cases that users may access data source detail in global.

Reply:
But can we add a unit test to cover the case when requestWorkspaceId is empty?